pandas/_libs/index.pyx

```diff
@@ -599,70 +600,72 @@ cpdef convert_scalar(ndarray arr, object value):
     return value
 
 
-cdef class MultiIndexObjectEngine(ObjectEngine):
-    """
-    provide the same interface as the MultiIndexEngine
-    but use the IndexEngine for computation
+cdef class BaseMultiIndexCodesEngine(object):
+    def __init__(self, levels, labels, offsets, **kwargs):
+        self._levels = levels
+        self._offsets = offsets
 
-    This provides good performance with samller MI's
-    """
-    def get_indexer(self, values):
-        # convert a MI to an ndarray
-        if hasattr(values, 'values'):
-            values = values.values
-        return super(MultiIndexObjectEngine, self).get_indexer(values)
+        # Map each combination to an integer
+        codes = (np.array(labels, dtype='int64').T + 1).astype('uint64')
+        lab_ints = self._codes_to_ints(codes)
 
-    cpdef get_loc(self, object val):
+        # Initialize underlying index
+        self._base.__init__(self, lambda: lab_ints, len(lab_ints), **kwargs)
```

Comment:
what is this for?

Reply:
"This" what? The fact that the entire engine is based on an uint engine? Or the fact that such integer engine must be initialized?

Reply:
what is ``self._base__`` it is not clear here

Reply:
It is the underlying engine, i.e. ``UInt64Engine`` or ``ObjectEngine``. I can't user ``super()`` because it would get ``__init__`` from the first superclass of (e.g.) ``MultiIndexUIntEngine``, which is ``BaseMultiIndexCodesEngine``, not ``UInt64Engine``. And I can't move ``UInt64Engine`` as first superclass, because I need ``MultiIndexUIntEngine.__init__`` to be ``BaseMultiIndexCodesEngine.__init__``. Don't know if there is a cleaner way... writing a different ``__init__`` for each subclass and then directly calling ``UInt64Engine.__init__`` is uglier.

Reply:
Well, except if I strip from ``BaseMultiIndexCodesEngine.__init__`` only the subclass specific portion, i.e.g

```python
class MultiIndexUIntEngine(...):
    def init_underlying(self, lab_ints):
        UInt64Engine.__init__(self, lambda: lab_ints, len(lab_ints)
```

I can do this if you think it's better.